Add ThrowingFunc family of Function interfaces.

mikebaum opened the pull request:
1. Added `ThrowingFunc{1 ... 9, N}` interfaces.
2. Added 2 static helper methods in `Async` for each throwing func to create an async function.
3. Added a utility class `ThrowingFunctions` with methods to convert each throwing function into a `Callable`.

This resolves #5.

mikebaum committed:
Add ThrowingFunc family of Function interfaces.
Update Async with static methods to convert throwing functions into async functions.

akarnokd commented on src/main/java/rx/util/async/Async.java:
```diff
@@ -15,19 +15,60 @@
+import rx.Scheduler;
 import rx.Scheduler.Worker;
-import rx.functions.*;
+import rx.Subscriber;
```
Could you make sure these remain as \* to so the header of the class won't show as merge conflict in case there are multiple PRs?

mikebaum replied:
No problem.

akarnokd commented on src/main/java/rx/util/async/Async.java:
```diff
@@ -92,17 +187,17 @@ private Async() {
      * @see <a href="http://msdn.microsoft.com/en-us/library/hh229182.aspx">MSDN: Observable.ToAsync</a>
      */
-    public static <R> Func0<Observable<R>> toAsync(Func0<? extends R> func) {
+    public static <R> Func0<Observable<R>> toAsync(Callable<R> func) {
```
This is a binary-incompatible API change.

mikebaum replied:
The thought was that `Func0` extends `Callable`, so I thought it should be fine. I could revert this change and add a second method `toAsyncThrowing(Callable)` if this does indeed cause a problem.

akarnokd commented on src/main/java/rx/util/async/Async.java:
```diff
@@ -500,40 +789,22 @@ private Async() {
      * @see <a href="http://msdn.microsoft.com/en-us/library/hh211792.aspx">MSDN: Observable.ToAsync</a>
      */
-    public static <R> Func0<Observable<R>> toAsync(final Func0<? extends R> func, final Scheduler scheduler) {
+    public static <R> Func0<Observable<R>> toAsync(final Callable<R> func, final Scheduler scheduler) {
```
Again, this is a binary-incompatible API change.

mikebaum replied:
Same as previous comment.

akarnokd commented on src/test/java/rx/util/async/AsyncTest.java:
```diff
@@ -819,6 +838,500 @@ public String answer(InvocationOnMock invocation) throws Throwable {
+        // Verifies that the returned Observable is "hot"
+        verify(func, times(1)).call();
+        
+        observable.subscribe(new TestObserver<Object>(observer));
```
Why do you wrap `observer` into a `TestObserver`? This goes for all the other places below.

mikebaum replied:
Good point, There is no need, I will remove the wrapping TestObserver.

mikebaum commented:
@akarnokd I updated the pull request addressing your review comments. Let me know if there is anything else. Also if you like I can issue another pull request with the three commits squashed if you prefer.

akarnokd commented:
Thanks for the update. It would be great if you squashed the commits in this PR. You should be able to do it without submitting a new PR, at least in Eclipse.

mikebaum commented:
After squashing the commits, I noticed that the signatures of `toAsync(Func0)` and `toAsync(Func0, Scheduler)` where still not the same, since originally they accepted `Func0<? extends R>`, not `Func0<R>`. I added a small commit to fix this issue. I chose not to squash that change, please check this change and then I will squash this change as well.

akarnokd commented:
Looks good to me.

mikebaum commented:
Done...

akarnokd commented:
:+1: 

Not sure when this can be merged because there is a rule that 2 :+1: is required for this type of change in RxJava. /cc @benjchristensen

mikebaum commented:
@akarnokd That sounds reasonable... Thanks for you're timely responses to this pull request.

mikebaum commented:
@akarnokd @benjchristensen Understandably, considering how busy you guys are, this pull request seems to have stalled. Any ideas when we can merge this? Is there anything else required?

akarnokd commented:
Thanks for the PR. I merged it but I can't really affect the release frequency of this library. Do you want to propose something else. If not, let's ask @abersnaze to make a release.

mikebaum commented:
@akarnokd Yeah, there's a couple of things:

1) I'd suggest refactoring the Async class to make all the calls to `toAsync` redirect to call `startCallable` since there is a lot of repetitive code in that class. Doing so would mean that wherever we call `toAsync` with a `Func` or an `Action` we should rather construct a Callable and then call `startCallable`. This could be done in place or using a helper class like I did with `ThrowingFunctions`.

2) Possibly update the api to return `Single` instead of `Observable` since as it is coded, we are only ever expecting one result.

I'd be willing to create issues and PRs for those changes.

Let m know what you think.

akarnokd commented:
1) I'm not very fond of wrapping but you can do a PR and we can see what it takes.

2) Changing the return type is not binary compatible so you'd have to create new methods. However, I'd like to improve things on `Single` directly if possible.
